File: Lecture033 Recursion Day3/binarySearch.py

def printa(arr, s, e):

    for i in range(s,e+1):
        print(arr[i],end=" ")
    print("\n")


def binarySearch( arr, s, e, k):
    print("\n")
    printa(arr, s, e)

    # base case

    # element not found
    if(s > e):
        return False

    mid = s + (e-s)//2
    print("value of arr mid is ",arr[mid])
    # element found
    if(arr[mid] == k):
        return True

    if(arr[mid] < k):
        return binarySearch(arr, mid+1, e, k)
    
    else:
        return binarySearch(arr, s, mid-1, k)
    



if __name__ == "__main__":

    arr = [2, 4, 6, 10, 14, 18, 22, 38, 49, 55, 222]
    size = 11
    key = 222

    print("Present or not ",binarySearch(arr, 0, size-1, key))

File: 033_Recursion_Day3/BinarySearch.java
public class BinarySearch {
    public static void main(String[] args) {
        int[] arr = {2 ,3 , 4 ,10 , 12, 122};
        int key = 3;

        int ans = findIdx(arr , 0 , 5  , key);

        if(ans == -1)
            System.out.println("Not Found");

        else
            System.out.println("Element present at index " + ans);
    }
    public static int findIdx(int[] arr , int s , int e , int key){
        //base case
        if(s > e)
            return -1;
        //find mid
        int mid = s + (e - s)/2;

        if(arr[mid] == key) // key is found
            return mid;

        if(arr[mid] < key) // search in right part
            return findIdx(arr , mid + 1 , e , key); // recurrence relation
        else // search in left part
            return findIdx(arr , s , mid - 1 , key);// recurrence relation
    }
}
